kotlin/reflect/src/main/resources/META-INF/com.android.tools/proguard/moshi-kotlin.pro

```diff
@@ -0,0 +1,26 @@
+# When editing this file, update the following files as well:
+# - META-INF/com.android.tools/r8-from-1.6.0/moshi-kotlin.pro
+# - META-INF/com.android.tools/r8-upto-1.6.0/moshi-kotlin.pro
+# - META-INF/proguard/moshi-kotlin.pro
+# Keep Metadata annotations so they can be parsed at runtime.
+-keep class kotlin.Metadata { *; }
+
+# Keep default constructor marker name for lookup in signatures.
+-keepnames class kotlin.jvm.internal.DefaultConstructorMarker
+
+# Keep names of Kotlin classes but allow shrinking if unused
+-keepnames @kotlin.Metadata class *
+
+# Keep fields, constructors, and methods in Kotlin classes.
+-keepclassmembers @kotlin.Metadata class * {
+  <fields>;
+  <init>(...);
+  <methods>;
+}
```

Comment:
@udalov @jsjeon I wonder if these should be added to the kotlin-reflect rules as well

Reply:
For some reason I thought that the `-keep class kotlin.Metadata` rule would also keep all members. If that's not true, then your proposal seems reasonable.

Reply:
This is slightly different, and rather it applies this rule only to classes annotated with Metadata, not the metadata annotation class itself

Reply:
thinking on it more, I think it's probably best to omit these for now and have the guidance be that users apply keep rules as needed for their exact cases. Same with kotlin-reflect - the rules in place now make sure _its_ machinery works, and users can just make sure they handle appropriate rules for classes that they plan to reflect on

Reply:
Cleaned up in ad47a56

Reply:
Oops, I should have paid attention to github alarms more often and carefully. Sorry for the late game.

@udalov 
> I thought that the -keep class kotlin.Metadata rule would also keep all members.

That's not true. The rule doesn't specify members, and thus all members are still eligible for shrinking and minification. Since members, such as `k`, `d1`, etc., are read all together (via metadata library), the correct rule is `-keep class kotlin.Metadata { *; }`.

But, as @ZacSweers already mentioned, rules here point to classes with metadata annotation, not the metadata definition itself.

Regarding having similar rules in kotlin-reflect, I would say no, because it's still up to users to add such rules if some classes (w/ or w/o metadata annotation) have reflective uses. Glad to see that you ended up with the same conclusion.